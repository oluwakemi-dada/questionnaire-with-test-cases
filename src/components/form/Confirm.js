import React from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import {
  DetailsWrapper,
  PrimaryHeading,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const Details = styled.div`
  background: rgba(195, 191, 243, 0.173);
  border-radius: 0.5rem;
  padding: 0 2rem;

  > *:last-child {
    border: 0;
  }
`;

const Items = styled.div`
  border-bottom: ${(props) =>
    props.theme === 'light' ? '1px solid #e4dbe7' : '1px solid #545353'};
  padding: 1rem 0;

  div {
    padding: 1rem 0rem;
    display: flex;
    align-items: center;
  }

  div:first-child {
    font-size: 1.4rem;
  }

  div:last-child {
    font-size: 1.55rem;
  }
`;

const Question = styled.div`
  color: ${(props) => (props.theme === 'light' ? '#565454' : '#b0aeae')};
`;

const Answer = styled.div`
  color: ${(props) => (props.theme === 'light' ? '#1c1c1c' : '#e8e5e5')};
`;

const Confirm = ({ prevStep, formData}) => {
  const history = useHistory();
  const currTheme = localStorage.getItem('theme');

  const { name, age, sex } = formData;

  // HANDLERS
  const proceed = (e) => {
    // e.preventDefault();
    history.push('/success');
  };

  const previous = (e) => {
    // e.preventDefault();
    prevStep();
  };

  return (
    <DetailsWrapper>
      <PrimaryHeading className='text-center '>Confirm</PrimaryHeading>
      <Details>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your name?</Question>
          <Answer theme={currTheme}>{name ? name : ''}</Answer>
        </Items>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your age bracket?</Question>
          <Answer theme={currTheme}>{age ? age : ''}</Answer>
        </Items>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your sex?</Question>
          <Answer theme={currTheme}>{sex ? sex : ''}</Answer>
        </Items>
      </Details>
      <BtnWrapper>
        <BtnPrev onClick={(e) => previous(e)}>Back</BtnPrev>
        <BtnNext onClick={(e) => proceed(e)}>Submit</BtnNext>
      </BtnWrapper>
    </DetailsWrapper>
  );
};

export default Confirm;
