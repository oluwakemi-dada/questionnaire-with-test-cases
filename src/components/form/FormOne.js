import React from 'react';
import styled from 'styled-components';
import Alert from '../layouts/Alert';
import { Wrapper, PrimaryHeading, FormControl, BtnNext } from './FormResources';

const BtnWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 4rem;
`;

const FormOne = ({ nextStep, onChange, formData, alert, setAlertFn }) => {
  const { name } = formData;

  // HANDLERS
  const proceed = () => {
    if (name === '') {
      setAlertFn('Name is required');
    } else {
      nextStep();
    }
  };
  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label htmlFor='name'>What is your name?</label>
        <input
          type='text'
          name='name'
          value={name}
          onChange={(e) => onChange(e)}
        />
      </FormControl>

      <BtnWrapper>
        <BtnNext onClick={proceed}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormOne;
