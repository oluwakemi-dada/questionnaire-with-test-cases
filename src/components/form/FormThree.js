import React from 'react';
import Alert from '../layouts/Alert';
import {
  Wrapper,
  PrimaryHeading,
  FormControl,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const FormThree = ({
  nextStep,
  prevStep,
  onChange,
  formData,
  alert,
  setAlertFn,
}) => {
  const { sex } = formData;

  // HANDLERS
  const proceed = (e) => {
    e.preventDefault();
    if (sex === '') {
      setAlertFn('Sex is required');
    } else {
      nextStep();
    }
  };

  const previous = (e) => {
    e.preventDefault();
    prevStep();
  };
  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label htmlFor='sex'>What is your sex?</label>
        <select name='sex' value={sex} onChange={(e) => onChange(e)}>
          <option value=''>Select ...</option>
          <option value='Male'>Male</option>
          <option value='Female'>Female</option>
          <option value='Decline to self-identity'>
            Decline to self-identity
          </option>
        </select>
      </FormControl>

      <BtnWrapper>
        <BtnPrev onClick={(e) => previous(e)}>Back</BtnPrev>
        <BtnNext onClick={(e) => proceed(e)}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormThree;
