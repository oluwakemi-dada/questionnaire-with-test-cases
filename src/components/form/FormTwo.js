import React from 'react';
import Alert from '../layouts/Alert';
import {
  Wrapper,
  PrimaryHeading,
  FormControl,
  Option,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const FormTwo = ({
  nextStep,
  prevStep,
  alert,
  setAlertFn,
  onChange,
  formData,
}) => {
  const { age } = formData;

  // HANDLERS
  const proceed = () => {
    if (age === '') {
      setAlertFn('Age is required');
    } else {
      nextStep();
    }
  };

  const previous = () => {
    prevStep();
  };

  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label>What is your age bracket?</label>

        <Option>
          <input
            name='age'
            type='radio'
            value='Below 18'
            checked={age === 'Below 18'}
            onChange={onChange}
          />
          <label htmlFor='age'>Below 18</label>
        </Option>
        <Option>
          <input
            name='age'
            type='radio'
            value='18 - 22'
            checked={age === '18 - 22'}
            onChange={onChange}
          />
          <label htmlFor='age'>18 - 22</label>
        </Option>
        <Option>
          <input
            name='age'
            type='radio'
            value='23 - 30'
            checked={age === '23 - 30'}
            onChange={onChange}
          />
          <label htmlFor='age'>23 - 30</label>
        </Option>
        <Option>
          <input
            name='age'
            type='radio'
            value='Above 30'
            checked={age === 'Above 30'}
            onChange={onChange}
          />
          <label htmlFor='age'>Above 30</label>
        </Option>
      </FormControl>

      <BtnWrapper>
        <BtnPrev onClick={previous}>Back</BtnPrev>
        <BtnNext onClick={proceed}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormTwo;
