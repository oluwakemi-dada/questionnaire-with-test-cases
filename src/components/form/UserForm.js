import React, { useState } from 'react';
import FormOne from './FormOne';
import FormTwo from './FormTwo';
import FormThree from './FormThree';
import Confirm from './Confirm';

const UserForm = () => {
  const [formData, setFormData] = useState({
    name: '',
    age: '',
    sex: '',
  });

  const [alert, setAlert] = useState('');

  const [step, setStep] = useState(1);

  // HANDLERS
  const onChange = (e) =>
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    });

  const setAlertFn = (msg) => {
    setAlert(msg);
    setTimeout(() => setAlert(''), 3000);
  };

  // Proceed to next step
  const nextStep = () => {
    setStep(step + 1);
  };

  // Go back to prev step
  const prevStep = () => {
    setStep(step - 1);
  };

  if (step === 1) {
    return (
      <FormOne
        nextStep={nextStep}
        onChange={onChange}
        formData={formData}
        alert={alert}
        setAlertFn={setAlertFn}
      />
    );
  } else if (step === 2) {
    return (
      <FormTwo
        prevStep={prevStep}
        nextStep={nextStep}
        onChange={onChange}
        formData={formData}
        alert={alert}
        setAlertFn={setAlertFn}
      />
    );
  } else if (step === 3) {
    return (
      <FormThree
        prevStep={prevStep}
        nextStep={nextStep}
        onChange={onChange}
        formData={formData}
        alert={alert}
        setAlertFn={setAlertFn}
      />
    );
  } else {
    return (
      <Confirm prevStep={prevStep} nextStep={nextStep} formData={formData} />
    );
  }
};

export default UserForm;
