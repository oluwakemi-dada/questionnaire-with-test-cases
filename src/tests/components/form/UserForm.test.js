import React from 'react';
import { shallow } from 'enzyme';
import UserForm from '../../../components/form/UserForm';
import { formData } from '../../fixtures/formData';
import FormOne from '../../../components/form/FormOne';
import FormTwo from '../../../components/form/FormTwo';
import FormThree from '../../../components/form/FormThree';
import Confirm from '../../../components/form/Confirm';

let wrapper;
beforeEach(() => {
  wrapper = shallow(<UserForm />);
});

test('should render FormOne when step = 1 (default)', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render FormTwo when step = 2', () => {
  wrapper.find(FormOne).prop('nextStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should render FormThree when step = 3', () => {
  wrapper.find(FormOne).prop('nextStep')();
  wrapper.find(FormTwo).prop('nextStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should render Confirm when step = 4', () => {
  wrapper.find(FormOne).prop('nextStep')();
  wrapper.find(FormTwo).prop('nextStep')();
  wrapper.find(FormThree).prop('nextStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should render FormThree when step = 4 and prevStep is called', () => {
  // Go to step 4 (Confirm)
  wrapper.find(FormOne).prop('nextStep')();
  wrapper.find(FormTwo).prop('nextStep')();
  wrapper.find(FormThree).prop('nextStep')();
  // prevStep should return to step 3 (FormThree)
  wrapper.find(Confirm).prop('prevStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should render FormTwo when step = 3 and prevStep is called', () => {
  // Go to step 3 (FormThree)
  wrapper.find(FormOne).prop('nextStep')();
  wrapper.find(FormTwo).prop('nextStep')();
  // prevStep should return to step 2 (FormTwo)
  wrapper.find(FormThree).prop('prevStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should render FormOne when step = 2 and prevStep is called', () => {
  // Go to step 2 (FormTwo)
  wrapper.find(FormOne).prop('nextStep')();
  // prevStep should return to step 1 (FormOne)
  wrapper.find(FormTwo).prop('prevStep')();

  expect(wrapper).toMatchSnapshot();
});

test('should update formData when onChange is called on FormOne', () => {
  const value = 'Oluwakemi Dada';
  const e = {
    target: {
      name: 'name',
      value,
    },
  };
  // Update
  wrapper.find(FormOne).prop('onChange')(e);
  // After update
  expect(wrapper.find(FormOne).prop('formData')).toEqual({
    ...formData,
    name: value,
  });
});

test('should update selectedOption when setSelectedOption is called on FormTwo', () => {
  const value = '23 - 30';
  wrapper.find(FormOne).prop('nextStep')();
  // Before update
  expect(wrapper.find(FormTwo).prop('selectedOption')).toBe('');
  // Update
  wrapper.find(FormTwo).prop('setSelectedOption')(value);
  // After update
  expect(wrapper.find(FormTwo).prop('selectedOption')).toBe(value);
});

test('should update formData when onChange is called on FormThree', () => {
  const value = 'Female';
  const e = {
    target: {
      name: 'sex',
      value,
    },
  };
  wrapper.find(FormOne).prop('nextStep')();
  wrapper.find(FormTwo).prop('nextStep')();
  // Before update
  expect(wrapper.find(FormThree).prop('formData')).toEqual({
    ...formData,
  });
  // Update
  wrapper.find(FormThree).prop('onChange')(e);
  // After update
  expect(wrapper.find(FormThree).prop('formData')).toEqual({
    ...formData,
    sex: value,
  });
});

test('should update alert when setAlertFn is called', () => {
  const msg = 'Name is required';
  // Before update
  expect(wrapper.find(FormOne).prop('alert')).toBe('');
  // Update
  wrapper.find(FormOne).prop('setAlertFn')(msg);
  // After update
  expect(wrapper.find(FormOne).prop('alert')).toBe(msg);
});
