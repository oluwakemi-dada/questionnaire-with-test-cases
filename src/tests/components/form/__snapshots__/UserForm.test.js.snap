// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Confirm when step = 4 1`] = `
<Confirm
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  prevStep={[Function]}
/>
`;

exports[`should render FormOne when step = 1 (default) 1`] = `
<FormOne
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  setAlertFn={[Function]}
/>
`;

exports[`should render FormOne when step = 2 and prevStep is called 1`] = `
<FormOne
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  setAlertFn={[Function]}
/>
`;

exports[`should render FormThree when step = 3 1`] = `
<FormThree
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  prevStep={[Function]}
  setAlertFn={[Function]}
/>
`;

exports[`should render FormThree when step = 4 and prevStep is called 1`] = `
<FormThree
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  prevStep={[Function]}
  setAlertFn={[Function]}
/>
`;

exports[`should render FormTwo when step = 2 1`] = `
<FormTwo
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  prevStep={[Function]}
  setAlertFn={[Function]}
/>
`;

exports[`should render FormTwo when step = 3 and prevStep is called 1`] = `
<FormTwo
  alert=""
  formData={
    Object {
      "age": "",
      "name": "",
      "sex": "",
    }
  }
  nextStep={[Function]}
  onChange={[Function]}
  prevStep={[Function]}
  setAlertFn={[Function]}
/>
`;
